warehouse/query-core/src/main/java/datawave/query/jexl/DatawavePartialInterpreter.java

```diff
@@ -0,0 +1,887 @@
+ * <ul>
+ * <li>(TRUE and TRUE) evaluates to TRUE and is a FULL evaluation</li>
+ * <li>(TRUE and FALSE) evaluates to FALSE and is a FULL evaluation</li>
+ * <li>(FALSE and FALSE) evaluates to FALSE and is a FULL evaluation</li>
+ * <li>(TRUE and UNKNOWN) evaluates to TRUE and is a PARTIAL evaluation</li>
+ * <li>(FALSE and UNKNOWN) evaluates to FALSE and is a FULL evaluation</li>
+ * <li>(UNKNOWN and UNKNOWN) evaluates to TRUE and is a PARTIAL evaluation</li>
+ * </ul>
+ * <h2>Evaluation Logic (Unions)</h2>
+ * <ul>
+ * <li>(TRUE or TRUE) evaluates to TRUE and is a FULL evaluation</li>
+ * <li>(TRUE or FALSE) evaluates to TRUE and is a FULL evaluation</li>
+ * <li>(FALSE or FALSE) evaluates to FALSE and is a FULL evaluation</li>
+ * <li>(TRUE or UNKNOWN) evaluates to TRUE and is a FULL evaluation</li>
+ * <li>(FALSE or UNKNOWN) evaluates to TRUE and is a PARTIAL evaluation</li>
+ * <li>(UNKNOWN or UNKNOWN) evaluates to TRUE and is a PARTIAL evaluation</li>
+ * </ul>
+ */
+public class DatawavePartialInterpreter extends DatawaveInterpreter {
+    
+    private static final Logger log = Logger.getLogger(DatawavePartialInterpreter.class);
+    private final Set<String> incompleteFields;
+    private final PartialInterpreterCallback callback;
+    
+    public DatawavePartialInterpreter(JexlEngine jexl, JexlContext aContext, boolean strictFlag, boolean silentFlag, Set<String> incompleteFields,
+                    PartialInterpreterCallback callback) {
+        super(jexl, aContext, strictFlag, silentFlag);
+        this.incompleteFields = incompleteFields;
+        this.callback = callback;
+        this.callback.reset();
+    }
+    
+    /**
+     * Sets the callback state if an incomplete field contributed to a positive evaluation
+     *
+     * @param node
+     *            the JexlNode
+     * @return true if the query matched the document
+     */
+    @Override
+    public Object interpret(JexlNode node) {
+        Object matched = super.interpret(node);
+        if (matched instanceof State) {
+            State state = (State) matched;
+            if (state.matched && state.fromUnknown) {
+                callback.setUsed(true);
+            }
+        } else {
+            throw new IllegalStateException("result of interpret was not STATE, it was " + matched.getClass().getSimpleName());
+        }
+        return matched;
+    }
+    
+    /**
+     * Returns true if the provided JexlNode contains an identifier that is also present in the set of incomplete fields
+     *
+     * @param node
+     *            the JexlNode
+     * @return true if incomplete fieldname found, false otherwise
+     */
+    private boolean isIncomplete(JexlNode node) {
+        for (String field : JexlASTHelper.getIdentifierNames(node)) {
+            field = JexlASTHelper.deconstructIdentifier(field);
+            // if the field is not in the context, we cannot evaluate it.
+            return incompleteFields.contains(field) && context.has(field);
```

Comment:
Yes, it is possible. This composite check of both the incompleteFields and the context allows us to appropriately defeat more documents on the tablet server. Consider the following case;

A query like `FOO == 'anchor' && filter:includeRegex(INCOMPLETE, 'value.*')` that produces a shard range means we could have a document of just `{FOO=anchor}`. In this case the absence of an incomplete field in the document means it will never evaluate to true, so we shouldn't even bother returning it to the webservice for a full evaluation.

The presence of the field in both the query and the set of incomplete fields is not enough to determine the state of the node in the interpreter.

Reply:
Another possible case would be `FIELD_A && (FIELD_B || INCOMPLETE)`. In this case of a document like `{FIELD_A, FIELD_B}` we can fully evaluate the query without having to run a secondary evaluation in the webservice. This is only possible if we check the context for the incomplete field.